components/brave_rewards/renderer/creator_detection_agent.cc

```diff
@@ -0,0 +1,265 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_rewards/renderer/creator_detection_agent.h"
+
+#include <array>
+#include <optional>
+#include <utility>
+
+#include "base/containers/fixed_flat_map.h"
+#include "base/feature_list.h"
+#include "base/functional/bind.h"
+#include "brave/components/brave_rewards/common/features.h"
+#include "brave/components/brave_rewards/common/publisher_utils.h"
+#include "brave/components/brave_rewards/resources/grit/creator_detection_generated.h"
+#include "content/public/renderer/render_frame.h"
+#include "gin/function_template.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+#include "third_party/blink/public/common/web_preferences/web_preferences.h"
+#include "third_party/blink/public/platform/platform.h"
+#include "third_party/blink/public/platform/scheduler/web_agent_group_scheduler.h"
+#include "third_party/blink/public/web/blink.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_script_source.h"
+
+namespace brave_rewards {
+
+namespace {
+
+constexpr auto kScriptMap = base::MakeFixedFlatMap<std::string_view, int>(
+    {{"github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"www.github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"gist.github.com", IDR_CREATOR_DETECTION_GITHUB_BUNDLE_JS},
+     {"reddit.com", IDR_CREATOR_DETECTION_REDDIT_BUNDLE_JS},
+     {"www.reddit.com", IDR_CREATOR_DETECTION_REDDIT_BUNDLE_JS},
+     {"twitch.tv", IDR_CREATOR_DETECTION_TWITCH_BUNDLE_JS},
+     {"www.twitch.tv", IDR_CREATOR_DETECTION_TWITCH_BUNDLE_JS},
+     {"twitter.com", IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS},
+     {"x.com", IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS},
+     {"vimeo.com", IDR_CREATOR_DETECTION_VIMEO_BUNDLE_JS},
+     {"www.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS},
+     {"m.youtube.com", IDR_CREATOR_DETECTION_YOUTUBE_BUNDLE_JS}});
+
+bool ShouldRunInIsolatedWorld(int resource_id) {
+  // By default, scripts are loaded into an isolated JS world. However, the
+  // Twitter script needs to access JS objects that are set directly on DOM
+  // nodes in the main world.
+  if (resource_id == IDR_CREATOR_DETECTION_TWITTER_BUNDLE_JS) {
+    return false;
+  }
+  return true;
+}
+
+v8::Isolate* GetIsolate(content::RenderFrame* render_frame) {
+  DCHECK(render_frame);
+  return render_frame->GetWebFrame()->GetAgentGroupScheduler()->Isolate();
+}
+
+struct DetectionScript {
+  std::string source;
+  bool should_run_in_isolated_world = true;
+};
+
+std::optional<DetectionScript> GetDetectionScript(
+    content::RenderFrame* render_frame) {
+  // Only run scripts for the main frame.
+  if (!render_frame || !render_frame->IsMainFrame()) {
+    return std::nullopt;
+  }
+
+  // Only run scripts if the user has enabled Brave Rewards.
+  if (!render_frame->GetBlinkPreferences().brave_rewards_enabled) {
+    return std::nullopt;
```

Comment:
are we making sure scripts are not injected in incognito contexts if rewards is enabled for the regular context?

Reply:
Yes - in [brave_content_renderer_client.cc](https://github.com/brave/brave-core/pull/25394/files#diff-e13db5dc9eb1cc4aa7a205b988a1ac2d04068e0fa9aa6cac27853aef0615e96dR155) the creator detection agent is not created for incognito contexts.